Improve taghints, github handler & add inline github issue search

jsmnbom opened the pull request:
**Taghints**
- Simply type `@roolsbot #` and be presented with a nice list of tag hints + their help
- Start typing, e.g. `@roolsbot #tuto` and it will suggest `#tutorial`

**Github Handler**
- Make github handler much muuuuch faster
- Instead of needing to download the entire webpage for an issue, we now just do a lightweight api request (and if part of `python-telegram-bot/python-telegram-bot`, it is already cached)

**Inline github**
- Github stuff like `#1230`, `bvanrijn/rules-bot#31`, `@9e5b84e2b698aba866f0816e3ee1c8c3320b188a`, `bvanrijn/rules-bot@600af3d3bd0665461d3d24f0b4f605719add0f57` now works in inline mode.
- You can also *search* for issues in `python-telegram-bot/python-telegram-bot` using`@roolsbot #SEARCHQUERY` and it will suggest a couple of issues. (note that `SEARCHQUERY` can't contain whitespace, but since it's doing fuzzy searching, omitting them usually works fine).
- It's possible to combine queries like `#10 #9`, `#10 #search` or even `#search1 bvanrijn/rules-bot#10 #search2`

jsmnbom committed:
Clean up inlinequery sending & make taghints easier to find

Simply type "#" to get a list of tags, and their help.

jsmnbom committed:
Make github lookups much faster and prepare for search

jsmnbom committed:
Better docs + add @ in front of sha

jsmnbom committed:
Clean up inlinequeries even more

- Only run the fuzzy code if the replacement char is in the query
- Only search docs/wiki if no other results found (as it just clutters the interface)

jsmnbom committed:
Fix default repo of github issues

jsmnbom committed:
Add auth to github api

Cause rate limit of 60 requests/hour is too danm low

jsmnbom committed:
Add inline github searching

jsmnbom committed:
Fix various things in github api

jh0ker commented on util.py:
```diff
@@ -58,3 +52,155 @@ def build_menu(buttons,
+        # No more issues
+        else:
+            # Add a job that every 10 min checks if the 100 first issues changed, and update them in our cache if needed
+            job_queue.run_repeating(lambda _, __: self._job(links['first']['url'],
```
Should this not be also a `run_once`? With `run_repeating`, won't it add a new additional job ever 10 minutes when this function is ran as part of the repeating job?

jsmnbom replied:
Yes, as confusing as it is to me, you're totally right :)

jh0ker commented on components/taghints.py:
```diff
@@ -98,27 +100,24 @@ def list_available_hints(bot, update):
                                         disable_web_page_preview=True)
 
 
-def get_hint_data(text):
```
This function is used in `taghints:128`

jsmnbom replied:
I had no idea taghints worked non-inline.. How did I miss that :/

jh0ker commented on components/inlinequeries.py:
```diff
@@ -1,116 +1,210 @@
+    return result_changed, result
+
+
+def inline_github(query):
```
I feel like this function needs some more work. I find it hard to understand what exactly is happening, despite the comments. I also get the following error on the inline query `@testbot #`

```
2018-10-04 10:58:06,593 - telegram.ext.dispatcher - ERROR - An uncaught error was raised while processing the update
Traceback (most recent call last):
  File "/home/jhoeke/git/rules-bot/venv/lib/python3.6/site-packages/telegram/ext/dispatcher.py", line 279, in process_update
    handler.handle_update(update, self)
  File "/home/jhoeke/git/rules-bot/venv/lib/python3.6/site-packages/telegram/ext/inlinequeryhandler.py", line 144, in handle_update
    return self.callback(dispatcher.bot, update, **optional_args)
  File "/home/jhoeke/git/rules-bot/components/inlinequeries.py", line 143, in inline_query
    results_list.extend(inline_github(query))
  File "/home/jhoeke/git/rules-bot/components/inlinequeries.py", line 100, in inline_github
    title = things[next(reversed(things))].title
StopIteration
```

jh0ker commented on components/inlinequeries.py:
```diff
@@ -1,116 +1,210 @@
+            commit = github_issues.get_commit(sha, owner, repo)
+            things[full] = commit
+
+    # If the last thing from the for loop above was a search
```
Why the last thing?

jh0ker commented on components/inlinequeries.py:
```diff
@@ -1,116 +1,210 @@
+    # If the last thing from the for loop above was a search
+    if search_query:
+        choices = []
+        # Output a separate choice for each search result
```
Why are you copying the whole dict? What else is in there?

jh0ker reviewed (changes requested):
I have some questions and some change requests. Taghints don't work anymore (due to deleted function I think). Some errors on the console as well.

jsmnbom committed:
Fix non-inline taghints

jsmnbom committed:
Fix scheduling of github issue updates

jsmnbom committed:
Try to explain inline_github better

jh0ker commented on util.py:
```diff
@@ -58,3 +52,156 @@ def build_menu(buttons,
+
+    def _get_json(self, url, data=None, headers=None):
+        # Add base_url if needed
+        url = url if url.startswith('https://') or url.startswith('http') else self.base_url + url
```
`url.startswith('https://') or url.startswith('http')` is redundant

jsmnbom committed:
Fix many things

- Allow searches that aren't the last thing in query
- Allow multiple searches (only last will give options, previous will just pick best one)
- Show titles in description
- Show last search result title as title if searching
- Fix replacement of #1 #10

jsmnbom committed:
Do actual error handling of issue getting

jsmnbom committed:
Fix RuntimeError when searching for issues

Happened because the issues dict would get updated while iterating over it

jsmnbom committed:
Fix error when bot name has underscore in it

jsmnbom committed:
Remove redundant url check

Code doesn't work with HTTP (non S) now, but itn't that just more secure ;-)

jh0ker reviewed (approved):
lets go
